load-gen/src/test/java/io/smartcat/data/loader/datasource/CSVDataSourceTest.java

```diff
@@ -0,0 +1,232 @@
+        // GIVEN
+        boolean result = false;
+        try (CSVParser csvParser = getCsvParser("simple-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+
+            // WHEN
+            result = dataSource.hasNext(0);
+        }
+
+        // THEN
+        Assert.assertEquals(true, result);
+    }
+
+    @Test
+    public void hasNext_should_return_false_at_the_beginning_of_empty_file() throws Exception {
+        // GIVEN
+        boolean result = false;
+        try (CSVParser csvParser = getCsvParser("empty-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+
+            // WHEN
+            result = dataSource.hasNext(0);
+        }
+
+        // THEN
+        Assert.assertEquals(false, result);
+    }
+
+    @Test
+    public void hasNext_should_return_false_at_the_end_of_file() throws Exception {
+        // GIVEN
+        boolean result = true;
+        try (CSVParser csvParser = getCsvParser("simple-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+            // file has NUM_OF_RECORDS_IN_SIMPLE_CSV records, run getNext() NUM_OF_RECORDS_IN_SIMPLE_CSV times and then
+            // try hasNext()
+            for (int i = 0; i < NUM_OF_RECORDS_IN_SIMPLE_CSV; i++) {
+                dataSource.getNext(0);
+            }
+
+            // WHEN
+            result = dataSource.hasNext(0);
+        }
+
+        // THEN
+        Assert.assertEquals(false, result);
+    }
+
+    @Test
+    public void first_getNext_should_return_first_row() throws Exception {
+        // GIVEN
+        User expected = new User("John", "Doe", 31, "555-232-13143", true);
+        User result = null;
+        try (CSVParser csvParser = getCsvParser("simple-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+
+            // WHEN
+            result = dataSource.getNext(0);
+        }
+
+        // THEN
+        Assert.assertEquals(expected, result);
+    }
+
+    @Test(expected = NoSuchElementException.class)
+    public void getNext_at_the_end_of_file_should_throw_NoSuchElementException() throws Exception {
+        // GIVEN
+        try (CSVParser csvParser = getCsvParser("simple-csv.csv");
+                CSVDataSource<User> dataSource = new CSVDataSource<>(csvParser, getRowMapper())) {
+            // file has NUM_OF_RECORDS_IN_SIMPLE_CSV records, run getNext() NUM_OF_RECORDS_IN_SIMPLE_CSV times and then
+            // try hasNext()
+            for (int i = 0; i < NUM_OF_RECORDS_IN_SIMPLE_CSV; i++) {
+                dataSource.getNext(0);
+            }
+
+            // WHEN
+            dataSource.getNext(0);
+        }
+
+        // THEN
+        // NoSuchElementException is thrown
+    }
+
+    @Test(expected = IllegalStateException.class)
```

Comment:
@nbozic Idea of testing exception message is a bit strange to me. Since it implies that exception message should be documented as part of the method contract. e.g. (@throws IllegalStateException with message "this and that") But I agree that this problem can happen. I would rather subclass IllegalStateException into something more appropriate. What do you think to subclass AlreadyClosedException from IllegalStateException?